File: app/src/main/java/com/imperium/academio/CustomUtil.java
package com.imperium.academio;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.os.Handler;
import android.util.Log;
import android.widget.EditText;

import com.google.android.material.textfield.TextInputLayout;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public final class CustomUtil {
    public static String SHA1(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-1");
            byte[] messageDigest = md.digest(input.getBytes());
            BigInteger no = new BigInteger(1, messageDigest);

            // Convert message digest into hex value
            StringBuilder hashText = new StringBuilder(no.toString(16));

            // Add preceding 0s to make it 32 bit
            while (hashText.length() < 32) {
                hashText.insert(0, "0");
            }
            return hashText.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    public static String validateField(TextInputLayout inputLayout, String type) {
        EditText editText = inputLayout.getEditText();
        if (editText == null)
            return null;
        String text = editText.getText().toString();
        String error = null;
        boolean valid = true;
        if (text.isEmpty()) {
            error = "Field cannot be empty";
        } else {
            switch (type) {
                case "username":
                    if (text.length() < 4 || text.length() > 15) {
                        error = "Username should have 4 to 15 characters";
                        valid = false;
                    }
                    if (valid && text.contains(" ")) {
                        error = "Username cannot contain spaces";
                        valid = false;
                    }
                    if (valid && !text.matches("^\\w{4,15}$")) {
                        error = "Username should be alphanumeric";
                    }
                    break;
                case "email":
                    String emailPattern = "[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+";
                    if (!text.matches(emailPattern)) {
                        error = "Email isn't valid";
                    }
                    break;
                case "password":
                    if (text.matches("^(.{0,7}|.{21,})$")) {
                        error = "Length of Password should be in range (8, 20)";
                        valid = false;
                    }
                    if (valid && text.matches("^([^A-Za-z]*)$")) {
                        error = "Password should contain at least one alphabet";
                        valid = false;
                    }
                    if (valid && text.matches("^([^0-9]*)$")) {
                        error = "Password should contain at least one digit";
                        valid = false;
                    }
                    if (valid && text.matches("^([a-zA-Z0-9]*)$")) {
                        error = "Password should contain at least one special character.";
                        valid = false;
                    }
                    if (valid && text.matches("[^\\s]*\\s.*")) {
                        error = "Password should not contain spaces.";
                    }
                    break;
            }
        }
        inputLayout.setError(error);
        if (error != null) { // Not valid so validator returns null
            new Handler().postDelayed(() -> {
                Context context = inputLayout.getContext();
                Activity activity = null;
                while (context instanceof ContextWrapper) {
                    if (context instanceof Activity) {
                        activity = (Activity) context;
                    }
                    context = ((ContextWrapper) context).getBaseContext();
                }

                if (activity != null)
                    activity.runOnUiThread(() -> {
                        inputLayout.setError(null);
                        inputLayout.setErrorEnabled(false);
                        Log.d("UIHERE", "validated");
                    });
            }, 10000);
            return null;
        } else { // No error so returns text
            inputLayout.setErrorEnabled(false);
            return text;
        }
    }
}

File: app/src/main/java/com/imperium/academio/CustomUtil.kt
package com.imperium.academio

import android.app.Activity
import android.content.ContextWrapper
import com.google.android.material.textfield.TextInputLayout
import java.math.BigInteger
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

object CustomUtil {
    @Suppress("FunctionName")
    @JvmStatic
    fun SHA1(input: String): String {
        try {
            val md = MessageDigest.getInstance("SHA-1")
            val messageDigest = md.digest(input.toByteArray())
            val no = BigInteger(1, messageDigest)

            // Convert message digest into hex value
            val hashText = StringBuilder(no.toString(16))

            // Add preceding 0s to make it 32 bit
            while (hashText.length < 32) {
                hashText.insert(0, "0")
            }
            return hashText.toString()
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException(e)
        }
    }

    @JvmStatic
    fun validateField(inputLayout: TextInputLayout, type: String?): String? {
        val editText = inputLayout.editText ?: return null
        val text = editText.text.toString()
        var error: String? = null
        var valid = true
        if (text.isEmpty()) {
            error = "Field cannot be empty"
        } else {
            when (type) {
                "username" -> {
                    if (text.length < 4 || text.length > 15) {
                        error = "Username should have 4 to 15 characters"
                        valid = false
                    }
                    if (valid && text.contains(" ")) {
                        error = "Username cannot contain spaces"
                        valid = false
                    }
                    if (valid && !text.matches("""^\w{4,15}$""".toRegex())) {
                        error = "Username should be alphanumeric"
                    }
                }
                "email" -> {
                    val emailPattern = """[a-zA-Z0-9._-]+@[a-z]+\.+[a-z]+""".toRegex()
                    if (!text.matches(emailPattern)) {
                        error = "Email isn't valid"
                    }
                }
                "password" -> {
                    if (text.matches("""^(.{0,7}|.{21,})$""".toRegex())) {
                        error = "Length of Password should be in range (8, 20)"
                        valid = false
                    }
                    if (valid && text.matches("""^([^A-Za-z]*)$""".toRegex())) {
                        error = "Password should contain at least one alphabet"
                        valid = false
                    }
                    if (valid && text.matches("""^([^0-9]*)$""".toRegex())) {
                        error = "Password should contain at least one digit"
                        valid = false
                    }
                    if (valid && text.matches("""^([a-zA-Z0-9]*)$""".toRegex())) {
                        error = "Password should contain at least one special character."
                        valid = false
                    }
                    if (valid && text.matches("""[^\s]*\s.*""".toRegex())) {
                        error = "Password should not contain spaces."
                    }
                }
                "year" -> if (!text.matches("""^[0-9]*$""".toRegex())) {
                    error = "Year should only contain digits."
                } else {
                    val year = text.toInt()
                    if (year < 1970 || year > 2038) {
                        error = "Year out of bounds."
                    }
                }
            }
        }
        inputLayout.error = error
        return if (error != null) { // Not valid so validator returns null
            inputLayout.postDelayed({
                var context = inputLayout.context
                var activity: Activity? = null
                while (context is ContextWrapper) {
                    if (context is Activity) {
                        activity = context
                    }
                    context = context.baseContext
                }
                activity?.runOnUiThread {
                    inputLayout.error = null
                    inputLayout.isErrorEnabled = false
                }
            }, 10000)
            null
        } else { // No error so returns text
            inputLayout.isErrorEnabled = false
            text
        }
    }
}